### DISCUSSION TITLE
Custom validations in MikroORM?

### DISCUSSION THREAD

**Discussion by Ashvith10** [mikro-orm/mikro-orm#6289] [Q&A] [answered]
I could validate fields in Active Record using [custom validations](https://guides.rubyonrails.org/active_record_validations.html#performing-custom-validations). I want to add a custom validation to check for collision of a string ID. How should I do the same in MikroORM?

**Comment by Ashvith10** [comment]
Soft ping to @B4nan and more expansion on the issue I am having:

I've created a simple helper function that looks something like this:
```js
import { EntitySchema, sql } from '@mikro-orm/core'
import { v4 } from 'uuid'
import { nanoid } from 'nanoid'
import { initORM } from '../plugins/mikro-orm.js'

/* -- Helper function generateCollisionFreeBase64Url -- */
async function generateCollisionFreeBase64Url() {
    const db = await initORM()
    while (true) {
        let base64Id = nanoid()
        if (db.url.findOne({ shortenedUrl: base64Id })) {
            return base64Id
        }
    }
}
/* -- End of helper function -- */

export class Url {
  constructor(originalUrl) {
    this.originalUrl = originalUrl
    this.shortenedUrl = generateBase64Url()
  }
}

export const urlSchema = new EntitySchema({
  class: Url,
  properties: {
    id: { type: 'uuid', onCreate: () => v4(), primary: true},
    originalUrl: { type: 'string', nullable: false },
    shortenedUrl: { type: 'string', nullable: false, unique: true },
    clicks: { type: 'number', nullable: false, default: 0 },
    date: { type: 'Date', default: sql.now() }
  }
})
```

The code syntax may, or may not be correct, because adding this inside the file that contains entity schema throws it into a recursion.

**Reply by B4nan** [reply]
That sounds way too convoluted, I would suggest leaving this up to the database instead, the insert query will fail if the generated ID is not unique, in such case just retry the process.

Or delegate this to a service method instead of trying it out inside a constructor. Your code cannot work, all those methods are async, you cant just do that inside a constructor and expect things to work.

Btw I wonder why you are even dealing with this, I don't think you should, the probability of collisions is extremely low, I am not sure if I ever cared about this. If this happens, users will get a bad request response, things are executed inside a transaction, so no harm done to the data, they will just retry the request.

**Reply by B4nan** [reply, marked as answer]
Or use hooks, `BeforeCreate` should be suitable for this and can be async.
